File: 0152-maximum-product-subarray/0152-maximum-product-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */

var maxProduct = function(nums) {
    let max = nums[0];
    let prevMin = nums[0];
    let prevMax = nums[0];
    
    for (let i = 1; i < nums.length; i++) {
        let currMax = Math.max(nums[i], nums[i]*prevMax, nums[i]*prevMin);
        let currMin = Math.min(nums[i], nums[i]*prevMax, nums[i]*prevMin);
        
        prevMax = currMax;
        prevMin = currMin;
        
        max = Math.max(max, currMax);
    }
    
    return max;
    
}


// Memoization
// var maxProduct = function(nums, memo=new Map()) {
//     // Base case
//     if (nums.length === 1) return nums[0];
    
//     if (nums in memo) return memo[nums];
//     memo[nums];
//     let max = nums[0];
    
//     // Compare product of prev & curr elem with curr elem 
//     let path1 = maxProduct(nums.slice(1), memo) 
//     nums[1] = nums[0]*nums[1];
//     let path2 = maxProduct(nums.slice(1), memo) // -2 * 3 = -6
    
//     max = Math.max(max, path1, path2);
//     // console.log('max: ' + max)
//     // memo[nums] = max;
//     // console.log(memo)
//     return max;
    
// };

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        prev_max = nums[0]
        prev_min = nums[0]
        max_prod = nums[0]
      
        i = 1
        while i < len(nums):
          curr_min = min(prev_min*nums[i], prev_max*nums[i], nums[i])
          curr_max = max(prev_min*nums[i], prev_max*nums[i], nums[i])
          prev_min = curr_min
          prev_max = curr_max
          max_prod = max(max_prod, prev_max)
          i += 1
        
        return max_prod